Libraries/LibWeb/FileAPI/FileReaderSync.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2025, Tim Ledbetter <tim.ledbetter@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibJS/Runtime/ArrayBuffer.h>
+#include <LibJS/Runtime/Promise.h>
+#include <LibJS/Runtime/Realm.h>
+#include <LibJS/Runtime/TypedArray.h>
+#include <LibTextCodec/Decoder.h>
+#include <LibWeb/Bindings/FileReaderSyncPrototype.h>
+#include <LibWeb/Bindings/Intrinsics.h>
+#include <LibWeb/FileAPI/Blob.h>
+#include <LibWeb/FileAPI/FileReaderSync.h>
+#include <LibWeb/HTML/EventLoop/EventLoop.h>
+#include <LibWeb/Platform/EventLoopPlugin.h>
+#include <LibWeb/Streams/AbstractOperations.h>
+#include <LibWeb/Streams/ReadableStream.h>
+#include <LibWeb/Streams/ReadableStreamDefaultReader.h>
+#include <LibWeb/WebIDL/ExceptionOr.h>
+
+namespace Web::FileAPI {
+
+GC_DEFINE_ALLOCATOR(FileReaderSync);
+
+FileReaderSync::~FileReaderSync() = default;
+
+FileReaderSync::FileReaderSync(JS::Realm& realm)
+    : PlatformObject(realm)
+{
+}
+
+void FileReaderSync::initialize(JS::Realm& realm)
+{
+    Base::initialize(realm);
+    WEB_SET_PROTOTYPE_FOR_INTERFACE(FileReaderSync);
+}
+
+GC::Ref<FileReaderSync> FileReaderSync::create(JS::Realm& realm)
+{
+    return realm.create<FileReaderSync>(realm);
+}
+
+GC::Ref<FileReaderSync> FileReaderSync::construct_impl(JS::Realm& realm)
+{
+    return FileReaderSync::create(realm);
+}
+
+// https://w3c.github.io/FileAPI/#dfn-readAsArrayBufferSync
+WebIDL::ExceptionOr<GC::Root<JS::ArrayBuffer>> FileReaderSync::read_as_array_buffer(Blob& blob)
+{
+    return read_as<GC::Root<JS::ArrayBuffer>>(blob, FileReader::Type::ArrayBuffer);
+}
+
+// https://w3c.github.io/FileAPI/#dfn-readAsBinaryStringSync
+WebIDL::ExceptionOr<String> FileReaderSync::read_as_binary_string(Blob& blob)
+{
+    return read_as<String>(blob, FileReader::Type::BinaryString);
+}
+
+// https://w3c.github.io/FileAPI/#dfn-readAsTextSync
+WebIDL::ExceptionOr<String> FileReaderSync::read_as_text(Blob& blob, Optional<String> const& encoding)
+{
+    return read_as<String>(blob, FileReader::Type::Text, encoding);
+}
+
+// https://w3c.github.io/FileAPI/#dfn-readAsDataURLSync
+WebIDL::ExceptionOr<String> FileReaderSync::read_as_data_url(Blob& blob)
+{
+    return read_as<String>(blob, FileReader::Type::DataURL);
+}
+
+template<typename Result>
+WebIDL::ExceptionOr<Result> FileReaderSync::read_as(Blob& blob, FileReader::Type type, Optional<String> const& encoding)
+{
+    // 1. Let stream be the result of calling get stream on blob.
+    auto stream = blob.get_stream();
+
+    // 2. Let reader be the result of getting a reader from stream.
+    auto reader = TRY(stream->get_a_reader());
+
+    // 3. Let promise be the result of reading all bytes from stream with reader.
+    auto promise_capability = reader->read_all_bytes_deprecated();
+
+    auto& promise = as<JS::Promise>(*promise_capability->promise());
+
+    // 4. Wait for promise to be fulfilled or rejected.
```

Comment:
It would be good for this to have a FIXME as well here like what we have in the `FileReader::read_operation` steps, as this spin is quite weird

Reply:
I've added both the relevant FIXMEs from `FileReader`.